### PR TITLE
Allow installation via package managers

### PR DESCRIPTION
**Pull request by jxa**
## Purpose and Motivation

SuperCollider currently bundles language modes for emacs, in addition to vim, gedit, etc. This is convenient for users installing from source, but this convenience comes with a cost. 
- The mode can not be installed independently without running cmake or manually editing the source code. 
- Updates to the source are more complicated because it requires updating scel plus bumping the supercollider submodule. 
- A number of recent issues (#8, #28, #30, #36) demonstrate that this is causing frustration.

Package managers for emacs have come a long way in the last 10+ years since the decision was made to bundle language modes with SC. Today emacs users have an expectation that they can install packages via a couple of standard methods, usually a package on [MELPA](https://github.com/melpa/melpa) or direct from a git repo via [straight.el](https://github.com/raxod502/straight.el).

## Changes

The strategy implemented here is to use native package managers for both emacs and supercollider. Once this PR is merged we can publish a package on melpa. Instructions are given for another popular emacs package manager. 

On the flip-side, the supercollider code will be distributed via Quarks. This PR adds a quark file and installation instructions.

We continue to support installation via `make`, though the mechanism had to change slightly. When compiling with cmake the `sclang-vars.el` file will be written, pointing to the proper data directory. When installed via package manager, the config it will guess the proper location by looking at the `system-type`, but it can be customized if it's not correct.

## Testing

I have verified the installation on MacOS (version 11.4) and on debian linux. On MacOS I followed the [new README instructions](https://github.com/supercollider/scel/pull/40/files), but substituted the following package definition because this PR isn't merged yet

```emacs-lisp
(package! sclang :recipe (:host github :repo "jxa/scel" :branch "packaging" :files ("el/*.el")))
```

On debian linux I installed supercollider from source, pointing the `editors/sc-el` submodule to this branch. Both halves of the code were installed properly and I was able to boot sclang and scserver from inside emacs (after requiring the library)

### CODE REVIEW

**Comment by jxa** [comment]
@dyfer and @joshpar this is now ready for review

**Comment by dyfer** [comment]
Thanks @jxa I'll try to test this soon.
Some comments questions so far:
- do I understand correctly that after merging this, we need 1) update sc-el in the main sc repo and 2) add this to MELPA? anything else we need?
- maybe the part in the README about MELPA should wait until we get the package in there? making that change later will be quick/easy to merge and it may avoid some confusion in the time frame between accepting this and getting the package into MELPA; OR we merge as is and give ourselves a short time frame to get the package is MELPA and if this doesn't happen in that time, you'd make another PR to make a note that the package is not yet at MELPA... what do you think?

**Comment by jxa** [comment]
Good point about the README @dyfer. I think it makes sense to pull out the bit about MELPA into a follow-up PR, which can be merged after the package is published. I'll make that change tonight.

> do I understand correctly that after merging this, we need 1) update sc-el in the main sc repo and 2) add this to MELPA? anything else we need?

As far as I know those are the only two

**Comment by jxa** [comment]
OK I pulled the documentation change out into #40.

**Review by bjornmossa — COMMENTED** [review]
Code looks good. Did you try to run it on vanilla Emacs?

**Comment by jxa** [comment]
> Code looks good. Did you try to run it on vanilla Emacs?

I have run it in a fresh debian VM, compiled from source, and in an existing install of emacs on my main machine (doom config), after uninstalling my existing setup. It's not currently possible to install with `package-install` because the package doesn't exist on melpa yet and it won't until this PR is merged.

**Comment by jxa** [comment]
@dyfer how do you feel about merging at this point? I commit to seeing this change through by publishing to melpa and getting the docs updated. If there are some bugs that come up on certain systems as a result of publishing I can work to iron those out too. 

Any risk of disruption on the compiled / linux side of things can be mitigated by delaying the supercollider/editors/scel submodule until we get some feedback (or spend some time with no bug reports)

Anything else I should be thinking about?

**Comment by dyfer** [comment]
@jxa I just want to test this myself before merging - should get to it sometime this week. Sorry for the delay!

**Comment by dyfer** [comment]
Hi @jxa 
Sorry - I'm not very familiar with Emacs and I'm having problems testing this. 
After I follow https://github.com/raxod502/straight.el to install `straight.el` (putting the boostrap code in my init file), when I try to run `(package! sclang :recipe (:host github :repo "jxa/scel" :branch "packaging" :files ("el/*.el")))` I am getting `Debugger entered--Lisp error: (void-function package!)`. Again, I haven't used Emacs (Aquamacs) in a very long time and I'm probably missing something basic... would you mind pointing me in the right direction here? Do I actually need https://github.com/raxod502/straight.el to install this package from your fork? 

As for readme, sorry for not being clear.
I think the readme should be a part of this PR, but for now it should not reference MELPA, but state how to install the package before it gets to MELPA, i.e. `(package! sclang :recipe (:host github :repo "supercollider/scel" :files ("el/*.el")))` or whatever the correct command it. After the package gets into MELPA, you would make another PR amending that section of the readme to include information about MELPA.

Looking at the readme in #40 I'd also suggest to start the section about the quark with the warning that it should not be installed if the sc-el mode was installed in a traditional manner (when building or using system package like `supercollider-emacs`.

**Comment by jxa** [comment]
Hi @dyfer I'm really glad you decided to check it out because it forced me to discover a new bug :)

The issue you have installing with straight.el is due to some bad instructions I provided. I'm using doom emacs, which has that `package!` wrapper function, but the way you can install without doom using straight.el is

```emacs-lisp
(straight-use-package
 '(sclang :type git :hots github :repo "jxa/scel" :branch "packaging" :files ("el/*.el")))
```

This is what exposes the bug. When installed this way, it looks like some files aren't being required properly so I get the error ` (void-function sclang-make-buffer-name)`. 

I have some time this weekend which I can use to sort out this new issue and update the readme. Thanks again, and sorry for wasting your time with imprecise instructions!

**Review by jxa — COMMENTED** [review]
A lot has changed, so you may want to review commit-by-commit. I tried to include good commit messages as I went.

I have tested this latest version in my personal config on mac os (doom config), a vanilla emacs install on mac os (straight.el config), and a fresh debian vm (compiled from supercollider source with this version in the editors/sc-el directory)

File: .github/workflows/test.yml
Comment (dyfer) [reply .github/workflows/test.yml:42]: Thanks for adding the CI!

**Comment by dyfer** [comment]
Thank you for your work on this and for detailed descriptions. I hope to get to this in the next few days, though it might take time as I don't know much about various aspects of scel. Your descriptions should make it easier though. GitHub Actions is definitely a nice "touch", thanks for adding that.

**Comment by jxa** [comment]
> it might take time as I don't know much about various aspects of scel

No problem @dyfer! If you would like a pair to help with installation and usage we could set up a time.

File: README.md
Comment (dyfer) [inline README.md:22]: Great! This works, BUT:
1. SuperCollider needs to be set to use the default `yaml` file, OR sc-el needs to be set to use the other/specified `yaml` file.
- SC IDE breaks after installing the quark (after sclang reboot) - we get duplicated classes

Re 1: IMO this needs to be spelled out in the README: when adding a quark, it's being added to the "compile directories" in the yaml file, the one that's specified in the SC IDE. The user either needs to use the default yaml file, or specify the desired yaml file on the sc-el side. It would be great to add a note somewhere (separate paragraph probably, referenced here maybe?) how to specify the yaml file to be used by sc-el

Re 2: When installing files with CMake, [they go into scide_scel folder](https://github.com/jxa/scel/blob/packaging/sc/CMakeLists.txt#L6). AFAIU the folder name is how IDEs recognize which classes to load/not load. In my testing it seems that renaming the folder `sc` to `scide_scel` in the scel repo fixes this. If that indeed is the solution, [this](https://github.com/jxa/scel/blob/packaging/CMakeLists.txt#L4) also needs to be updated accordingly and installing with cmake needs to be tested again.
BTW would it be too much to add installation with CMake to GHA? That's kind of involved since you'd need to install whole SC first), but maybe would be a good addition? That's rather optional, and could be easily added later. But fixing the issue with duplicated classes in SC-IDE definitely needs to be addressed.

File: README.md
Comment (dyfer) [inline README.md:20]: Writing from an emacs newbie perspective (well, I forgot everything I've done with it years ago), it might be beneficial to indicate that these go into the init/preferences file, instead of being invoked once.

File: README.md
Comment (dyfer) [inline README.md:29]: Writing from an emacs newbie perspective (well, I forgot everything I've done with it years ago), it might be beneficial to indicate that these go into the init/preferences file, instead of being invoked once. (Similarly for the "manual" install and adding sclang to the path on macOS)

File: README.md
Comment (dyfer) [inline README.md:41]: minor thing - `MacOS` -> `macOS`

File: README.md
Comment (dyfer) [inline README.md:52]: Maybe the fact that we have two ways of installing scel should be elaborated a little more - possibly not in this section though, but in the Introduction. Also I don't think that CMake install technically Linux-only (thought I might be wrong). But yes, it is the current way to get it on Linux, so I think we could say that it could be installed using either method - when compiling, OR using the new method (but not both! - important to spell that out)

File: el/Eldev
Comment (dyfer) [reply el/Eldev:7]: Pardon my vague understanding: is `eldev` being installed with scel? It is always available in emacs? Also, what's the purpose of the commented out lines above?

File: README.md
Comment (dyfer) [inline README.md:72]: Again, newbie comment: we need to be in `sclang-mode` first, so that needs to be activated for new buffers.

**Review by dyfer — CHANGES_REQUESTED** [review]
Thanks!
Here's a "first round" of comments from trying your PR on macOS with Aquamacs. The issue with duplicate class names with SC-IDE is the biggest issue, and probably a change to fix it (renaming a folder) needs to be carefully checked for regressions with the CMake installation mode. More below.

File: README.md
Comment (dyfer) [reply README.md:22]: After thinking a little more about this, adding Cmake install (and SC install) to GHA should not be part of this PR - to be possibly considered later. On thing to consider would be to run such test on the main SC repo, where we already build SC.

File: README.md
Comment (dyfer) [reply README.md:52]: Also, debian/ubuntu (maybe other distros too?) provide supercollider-emacs package. It might be worth noting that CMake install is equivalent to the system package install and should not be performed together with the "new" install method.

File: README.md
Comment (jxa) [reply README.md:22]: I see there is a variable `sclang-library-configuration-file` that can be set by the emacs user if they want to customize the yaml file to load. I can add some text to the readme to clarify this.

It took me a little while to understand what was happening in (2), but your suggestion appears to work. I'm close to shipping that change, I just need to verify that it still works when compiling from source. (which I need to do in a VM because I'm on a M1 macbook :-/ )

File: el/Eldev
Comment (jxa) [reply el/Eldev:6]: No it's not a built-in. Removed those commented-out lines and replaced with a better explanation of what this file is about and how to obtain eldev.

File: README.md
Comment (jxa) [reply README.md]: I was able to confirm that after renaming the `sc` directory, installation from source works in both Emacs and SC IDE with no errors.

**Comment by rukano** [comment]
Hi, I finally got to test this, and got my emacs-macs with doom running with SC!  Thanks for this!
I'm just getting one error when I start the interpreter:

```
*** Welcome to SuperCollider 3.11.2. *** For help type C-c C-y.
ERROR: syntax error, unexpected '(', expecting NAME or CLASSNAME or WHILE or '['
  in interpreted text
  line 1 char 69:

  Emacs.lispPerformCommand('Documentsetproperty', [ 2, 'prSetTitle', #("*doom*" 0 6 (face nil)) ], false) 
  ```
  
  But doesn't seem to affect the rest of the functionality :)

File: el/sclang-help.el
Comment (dyfer) [reply el/sclang-help.el:26]: Do you mean that `'ignmore-missing-file` makes this check optional?

File: README.md
Comment (dyfer) [inline README.md:83]: If we make the installing from source deprecated/obsolete, we should indicate it in the readme. Probably for later, not in this PR, but just mentioning it for the future.

File: el/sclang-help.el
Comment (jxa) [reply el/sclang-help.el:26]: Yes, normally `(require 'a-missing-package)` would raise a file-missing error. The "truthy" value for the third argument just suppresses that error. `sclang-vars` will be present when compiled with cmake. It will be absent when installed from github / pkg manager.

File: README.md
Comment (jxa) [reply README.md:83]: I had a similar thought, but I didn't want to presume too much. Perhaps after we are listed on MELPA?

**Comment by jxa** [comment]
> ```
> ERROR: syntax error, unexpected '(', expecting NAME or CLASSNAME or WHILE or '['
>   in interpreted text
>   line 1 char 69:
> 
>   Emacs.lispPerformCommand('Documentsetproperty', [ 2, 'prSetTitle', #("*doom*" 0 6 (face nil)) ], false) 
> ```

@rukano thanks for trying it out! I'm curious exactly how to recreate that error. Can you share the relevant part of your doom config and what command you execute to cause that error?

**Comment by dyfer** [comment]
Thanks @jxa 
Couple more thoughts since I reviewed it:
- I'm wondering whether it's maybe more elegant to keep the supercollider code in `sc` top-level directory, what do you think? If we were to do that, we could move the SC code into `sc/scide_scel`, what do you think about that? 
- I'd still would like to see a sentence in the README like: "If you open a new buffer in emacs, run `M-x sclang-mode` to activate sclang mode" or something like that. Unless I'm missing something, this is not already present, correct?
- This is a possible larger question, but let's think about it for a second: if we install the quark, couldn't we use that repository instead of installing a package through `straight-use-package` or MELPA? Is that awkward in emacs ecosystem? Do I understand correctly that it could look like this:
```supercollider
// run in sclang
Quarks.install("....scel");
```
```lisp
;; in emacs, add to the ~/.emacs file
(add-to-list 'load-path "~/Library/Application Support/SuperCollider/download-quarks/scel/el/") ;; check the path and provide examples for other OSs as well
(require 'sclang)
```

I don't want to de-rail the PR, it's already such an improvement, but this came to my mind recently and I wanted to get your opinion on it.

**Comment by jxa** [comment]
>     * I'm wondering whether it's maybe more elegant to keep the supercollider code in `sc` top-level directory...? If we were to do that, we could move the SC code into `sc/scide_scel`, what do you think about that?

I like it. I was a little sad to lose the `/sc` and `/el` symmetry :) I'll see if that works

>     * I'd still would like to see a sentence in the README like: "If you open a new buffer in emacs, run `M-x sclang-mode` to activate sclang mode" or something like that. Unless I'm missing something, this is not already present, correct?

`sclang-mode` is automatically activated when a `.sc` or `.scd` file is opened. Then all that is necessary is to call `sclang-start` (bound to `C-c C-o`).

One thing that bugs me: I would like to be able to autoload the `sclang-start` function (be able to call it before you `(require 'sclang)`). I haven't been able to get that working yet, so for now you either have to have a `scd` file or call `require` first. 

>     * if we install the quark, couldn't we use that repository instead of installing a package through `straight-use-package` or MELPA? Is that awkward in emacs ecosystem? 

It's an excellent question. Something I hadn't considered. It would simplify the install instructions at least :) 

It's a good idea and it wouldn't change anything for users who prefer to use straight, doom, etc. I was reviewing MELPA submission guidelines recently and I think it will take some more work to get this library to a place where they would list it. There are lots of compilation warnings, for example, which I haven't addressed because I felt it would be better to have a different PR for it.

So I'll rework the install instructions to remove reference to emacs package managers and have them add `Quarks.folder` to their load path and require from there. Sound good?

**Comment by dyfer** [comment]
> I like it. I was a little sad to lose the `/sc` and `/el` symmetry :) I'll see if that works

Great!

> `sclang-mode` is automatically activated when a `.sc` or `.scd` file is opened. Then all that is necessary is to call `sclang-start` (bound to `C-c C-o`).

Yes, but what if you are starting with a new (empty) document? This is what I was doing when testing this, and it took me a moment to realize I need to activate sclang-mode. Or is the user supposed to open a new document in the sclang-mode from the get-go? Not sure I saw an option for that in Aquamacs's menu.

> One thing that bugs me: I would like to be able to autoload the `sclang-start` function (be able to call it before you `(require 'sclang)`). I haven't been able to get that working yet, so for now you either have to have a `scd` file or call `require` first.

It sounds like a worthy addition, but I can't help/comment more on that with my limited knowledge of emacs/lisp, sorry.

> It's a good idea and it wouldn't change anything for users who prefer to use straight, doom, etc. I was reviewing MELPA submission guidelines recently and I think it will take some more work to get this library to a place where they would list it. There are lots of compilation warnings, for example, which I haven't addressed because I felt it would be better to have a different PR for it.

Good to know, thanks for looking into that.

> So I'll rework the install instructions to remove reference to emacs package managers and have them add `Quarks.folder` to their load path and require from there. Sound good?

Yes, that sounds great! Streamlining the installation instructions is definitely a plus!

One more comment for the [readme](https://github.com/supercollider/scel/pull/39/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R13): I'd make a warning stand out, something like: 
> Whatever option you choose, make sure not to mix installation methods. In particular, **do not install the quark** if you already have the `supercollider-emacs` package or if you compiled SuperCollider with the `-DSC_EL=ON` option.

Because SC "breaks" (library compilation fails) with duplicated classes, IMO it is worthy to make this clear.

**Comment by jxa** [comment]
All the readme comments I think are now addressed. I also figured out how to autoload `sclang-start` and added instructions for that. Thanks for all the feedback! We're in a much better place as a result

File: el/sclang-help.el
Comment (jxa) [reply el/sclang-help.el:44]: I'd prefer to leave this out of this PR because
- I don't know of anyone who could test this
- It can easily be set via `customize`

Hopefully some day a windows user will send us a PR :)

File: el/sclang-help.el
Comment (jxa) [reply el/sclang-help.el:52]: I also find this a bit strange. It adds in the customizable `sclang-system-help-dir` with this static, linux-specific, path. So it's possible to get the correct behavior via `customize`.

It does work for me on mac so I didn't want to mess with it. In the future I can look to clean it up.

File: README.md
Comment (jxa) [reply README.md]: It is not required, but the w3m functionality does still work. I changed the text to try to clarify that this is optional. Does it make more sense now?

File: README.md
Comment (jxa) [reply README.md]: I ended up using your text.

I was able to set the value by selecting `1 file` from the menu and then tabbing over and filling in the file field with the filename. Screenshot below

Screen Shot 2021-09-04 at 9 16 59 AM

In my opinion this experience is a bit clunky and I think it can be improved by changing the `defcustom` options. If it's ok with you I'd prefer to do that in a separate PR.

File: README.md
Comment (jxa) [reply README.md:89]: > I've tried installing on Ubuntu and it works. However, at first it didn't - it seems that Emacs would not load scel from /usr/local/share/emacs.... After copying these files to /usr/share/emacs... everything worked. I'm not sure if this is particular to my system (Ubuntu 20.04) or whether this is something to mention in the readme as well.

To this section I have added the snippet of code I put in my .emacs file on ubuntu, which I think will solve your issue and you won't need to copy the files

File: README.md
Comment (dyfer) [reply README.md]: Perfect, thanks so much!

> I was able to set the value by selecting 1 file from the menu and then tabbing over and filling in the file field with the filename. Screenshot below

That sounds great, I thought this might be something on my end so if you were able to set this, we're all set.

> In my opinion this experience is a bit clunky and I think it can be improved by changing the `defcustom` options. If it's ok with you I'd prefer to do that in a separate PR.

I agree on both accounts - clunky experience, for another PR.

File: README.md
Comment (dyfer) [reply README.md]: I didn't realize that. Thanks for adding the explanation, now it's clear.

**Review by dyfer — APPROVED** [review]

**Comment by dyfer** [comment]
I've made a minor change to the readme, but I somehow messed up the first commit, sorry! I'm merging this now. Thanks again for your work on this!

**Comment by friskgit** [comment]
I was very happy to see that this work is going on. I have completely missed it. I will try it as soon as I can! One question though: is this mainly a new method for installation, or are there changes to the sclang emacs mode itself?

**Comment by jxa** [comment]
> is this mainly a new method for installation, or are there changes to the sclang emacs mode itself?

Hi @friskgit, this PR focused on methods of installation only. Some changes were made to how the code is organized (to enable proper autoloading) and one custom variable was added.

**Pull request merged**
